clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/PersistenceDAO.scala

```diff
@@ -78,6 +100,84 @@ trait PersistenceDAO extends LazyLogging {
       )
       logger.debug(s"Wrote document $document to ${written.toUri}")
     }
+
+  /**
+    * Throw or return the `ClioDocument` at `path`.
+    * @param path to a JSON-encoded `ClioDocument``
+    * @tparam D is the type of `ClioDocument` at `path`
+    * @return the `ClioDocument` at `path`
+    */
+  def documentFromPath[D <: ClioDocument: Decoder](path: Path): D = {
+    decode[D](new String(Files.readAllBytes(path))).fold(throw _, identity)
+  }
+
+  /**
+    * @param path in some params `FileSystem`
+    * @return true if `path` is a "params file"
+    */
+  def isParamsFile(path: Path): Boolean = {
+    val s = path.toString
+    val parts = s.split(File.separator)
+    val last = parts.last
+    val isJson = last.endsWith(".json")
+    val hasUuid = last.split('-').length == 5
+    parts.length == 6 && isJson && hasUuid
+  }
+
+  /**
+    * Return the `ClioDocument`s under `rootdir` upserted no earlier
+    * than `withId`.
+    *
+    * @param rootDir is path to a bucket of "params files"
+    * @param withId is the path of earliest "params file" to return
+    * @param ec is an `ExecutionContext`
+    * @tparam D is the type of `ClioDocument` expected under `rootDir`
+    * @return the `ClioDocument` at `withId` and all later ones
+    */
+  def sinceId[D <: ClioDocument: Decoder](rootDir: Path, withId: Path)(
+    implicit ec: ExecutionContext
+  ): Future[Seq[D]] = {
+    val before = (p: Path) => p.compareTo(withId) < 0
+    val after = Directory
+      .walk(rootDir)
+      .filter(p => isParamsFile(p))
+      .filter(!before(_))
```

Comment:
Can we move the `.map(p => documentFromPath(p))` up to be right under this line, then sort by upsertId directly? Then I think we should only need one `.runWith`

Reply:
Yep.